pinot-tools/src/main/java/org/apache/pinot/tools/PinotDataAndQueryAnonymizer.java

```diff
@@ -530,68 +544,107 @@ private void generateDerivedValuesForGlobalDictionary(OrigAndDerivedValueHolder
+        valueHolder.setDerivedValues(generateDerivedIntValuesForGD(cardinality));
         break;
       case LONG:
-        valueHolder.setDerivedValues(generateDerivedLongValues(cardinality));
+        valueHolder.setDerivedValues(generateDerivedLongValuesForGD(cardinality));
         break;
       case FLOAT:
-        valueHolder.setDerivedValues(generateDerivedFloatValues(cardinality));
+        valueHolder.setDerivedValues(generateDerivedFloatValuesForGD(cardinality));
         break;
       case DOUBLE:
-        valueHolder.setDerivedValues(generateDerivedDoubleValues(cardinality));
+        valueHolder.setDerivedValues(generateDerivedDoubleValuesForGD(cardinality));
         break;
       case STRING:
-        valueHolder.setDerivedValues(generateDerivedStringValues(valueHolder));
+        valueHolder.setDerivedValues(generateDerivedStringValuesForGD(valueHolder));
+        break;
+      case BYTES:
+        valueHolder.setDerivedValues(generateDerivedByteValuesForGD(valueHolder));
         break;
       default:
         throw new UnsupportedOperationException("global dictionary currently does not support: " + valueHolder._dataType.name());
     }
   }
 
-  private Integer[] generateDerivedIntValues(int cardinality) {
+  private Integer[] generateDerivedIntValuesForGD(int cardinality) {
     Integer[] values = new Integer[cardinality];
     for (int i = 0; i < cardinality; i++) {
       values[i] = INT_BASE_VALUE + i;
     }
     return values;
   }
 
-  private Long[] generateDerivedLongValues(int cardinality) {
+  private Long[] generateDerivedLongValuesForGD(int cardinality) {
     Long[] values = new Long[cardinality];
     for (int i = 0; i < cardinality; i++) {
       values[i] = LONG_BASE_VALUE + (long)i;
     }
     return values;
   }
 
-  private Float[] generateDerivedFloatValues(int cardinality) {
+  private Float[] generateDerivedFloatValuesForGD(int cardinality) {
     Float[] values = new Float[cardinality];
     for (int i = 0; i < cardinality; i++) {
       values[i] = FLOAT_BASE_VALUE + (float)i;
     }
     return values;
   }
 
-  private Double[] generateDerivedDoubleValues(int cardinality) {
+  private Double[] generateDerivedDoubleValuesForGD(int cardinality) {
     Double[] values = new Double[cardinality];
     for (int i = 0; i < cardinality; i++) {
       values[i] = DOUBLE_BASE_VALUE + (double)i;
     }
     return values;
   }
 
-  private String[] generateDerivedStringValues(OrigAndDerivedValueHolder valueHolder) {
+  private String[] generateDerivedStringValuesForGD(OrigAndDerivedValueHolder valueHolder) {
     int cardinality = valueHolder._index;
     String[] values = new String[cardinality];
     for (int i = 0; i < cardinality; i++) {
-      int origValLength = ((String)valueHolder._origValues[i]).length();
+      String val = (String)valueHolder._origValues[i];
+      int origValLength;
+      if (val == null || val.equals("") || val.equals(" ") || val.equals("null")) {
```

Comment:
`if (val == null || val.equals("null)) {
  // choose "null" as the GD value
}`

Reply:
done